00_Base/src/interfaces/repository.ts

```diff
@@ -61,46 +134,108 @@ export interface ICrudRepository<T> {
    * @param namespace - The namespace in which to update the key.
    * @returns A promise that resolves to the updated value, or undefined if the key does not exist.
    */
-  updateByKey?(
-    value: T,
+  public async updateByKey(
+    value: Partial<T>,
+    key: string,
+    namespace?: string,
+  ): Promise<T | undefined> {
+    const result = await this._updateByKey(value, key, namespace);
+    this.emit('updated', result ? [result] : []);
+    return result;
+  }
+  protected abstract _updateByKey(
+    value: Partial<T>,
     key: string,
     namespace?: string,
   ): Promise<T | undefined>;
 
   /**
-   * Updates the value associated with the given query in the specified namespace.
+   * Updates the values associated with the given query in the specified namespace.
    * If no namespace is provided, the default namespace is used.
    *
    * @param value - The new value to associate with the query.
    * @param query - The query to use.
-   * @param namespace - Optional namespace for the kqueryey.
-   * @returns A promise that resolves to the value associated with the key, or undefined if the key does not exist.
+   * @param namespace - Optional namespace for the query.
+   * @returns A promise that resolves to the updated values associated with the query.
    */
-  updateByQuery(
-    value: T,
+  public async updateAllByQuery(
+    value: Partial<T>,
     query: object,
     namespace?: string,
-  ): Promise<T | undefined>;
+  ): Promise<T[]> {
+    const result = await this._updateAllByQuery(value, query, namespace);
+    this.emit('updated', result);
```

Comment:
should we also emit the Class as well as the object or some sort of type along or do you think the object itself will be sufficient? My thought process here is looking at some utility behind redux and when the event is emitted with just the object, it means that you need to have a good parser to have multiple handlers for the different classes that are being emitted. I think you can pass the Class itself along as well (similar to how we do it in java like ClassName.class but with different syntax) - and I think if you pass class along then you can just have a simple switch down the line to parse the emitted event and be able to quickly pass the payload to the correct handler.

It is kind of hard to explain, so I will be happy to clarify!

Reply:
There's no worry of any of this. The typing is supplied from the function signature.